In [16]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

In [8]:
df = pd.read_csv("house_price_regression_dataset.csv")

df.head()

,Square_Footage,Num_Bedrooms,Num_Bathrooms,Year_Built,Lot_Size,Garage_Size,Neighborhood_Quality,House_Price
0,1360,2,1,1981,0.599637,0,5,2.623829e+05
1,4272,3,3,2016,4.753014,1,6,9.852609e+05
2,3592,1,2,2016,3.634823,0,9,7.779774e+05
3,966,1,2,1977,2.730667,1,8,2.296989e+05
4,4926,2,1,1993,4.699073,0,8,1.041741e+06


In [9]:
X = df.drop("House_Price", axis=1)
Y = df["House_Price"]

In [10]:
X_train, X_test, Y_train, Y_test = train_test_split(
    X, Y, test_size=0.2, random_state=42)

Step 1: Train a LinearRegression model on a provided regression dataset (e.g. house prices).

In [12]:
model = LinearRegression()
model.fit(X_train, Y_train)
predictions = model.predict(X_test)

Step 2: Report the model's coefficients and identify which feature has the strongest effect.

In [15]:
print(model.coef_)
print(model.intercept_)

[  199.5132451  10225.20442447  8208.43477773   993.53717106
 14885.38441462  5146.14838284   115.06859524]
-2010567.4202702092


the strongest coefficient is the largest one, Lot_Size has the strongest effect since it has a largest coefficient (14885.38441462), it means that one-unit increase in Lot_Size is the highest increase in House_Price.
but in fact I want to add a note from me, it is not mean that it is the strongest feature, 'cause the numbers in Lot_Size is too small if we compare it for example with Square_Footage.

Step 3: Evaluate the model with MAE, RMSE, and R² on the test set.

In [22]:
print("MAE =", mean_absolute_error(Y_test, predictions))
print("RMSE =", np.sqrt(mean_squared_error(Y_test, predictions)))
print("R² =", r2_score(Y_test, predictions))

MAE = 8174.583600008711
RMSE = 10071.484424138673
R² = 0.9984263636823408


Step 4: Compare the RMSE against a baseline that predicts the mean for every row, and state whether the 
model adds value.

In [24]:
baseline_pred = np.full_like(Y_test, Y_train.mean(), dtype=float)
baseline_rmse = np.sqrt(mean_squared_error(Y_test, baseline_pred))
print(baseline_rmse)
print(np.sqrt(mean_squared_error(Y_test, predictions)))

253891.4870598681
10071.484424138673


Baseline RMSE (predicting the mean for every row) = 253891.4870598681, and the model RMSE = 10071.484424138673. the model RMSE is smaller than the baseline, and R2 = 0.998, this means that the model explains 99.8% of the variance in house price. it is very good.

Step 5: Document the interpretation of results in Markdown.

The Regression model was trained to predict house price using 7 features:
Square_Footage	Num_Bedrooms	Num_Bathrooms	Year_Built	Lot_Size	Garage_Size	Neighborhood_Quality.



the coefficient of Lot_Size has the strongest effect since it has a largest coefficient (14885.38441462), it means that one-unit increase in Lot_Size is the highest increase in House_Price. but it not mean that the Lot_Size has a strongest effest on the model, 'cause the features have different scale.

in the test set, the model have thids results MAE = 8174.583, RMSE = 10071.484, R² = 0.998, which mens that the model explains 99.8% of variance of the house price.

Baseline RMSE (predicting the mean for every row) = 253891.487, and the model RMSE = 10071.484424138673. the model RMSE is smaller about 25X than the baseline. this show us thst the model has learned meaningful patterns.

conclution: the model performs very well on this dataset with low error.